File: src/prod/src/Management/ImageModel/RunLayoutSpecification.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace std;
using namespace Common;
using namespace Management;
using namespace ImageModel;

GlobalWString RunLayoutSpecification::SharedFolderSuffix = make_global<wstring>(L".lk");
RunLayoutSpecification::RunLayoutSpecification()
    : runRoot_()
{
}

RunLayoutSpecification::RunLayoutSpecification(wstring const & runRoot)
    : runRoot_(runRoot)
{
}

wstring RunLayoutSpecification::GetApplicationFolder(
    wstring const & applicationId) const
{
    return Path::Combine(runRoot_, applicationId);
}

wstring RunLayoutSpecification::GetApplicationWorkFolder(
    wstring const & applicationId) const
{
    return Path::Combine(GetApplicationFolder(applicationId), L"work");
}

wstring RunLayoutSpecification::GetApplicationLogFolder(
    wstring const & applicationId) const
{
    return Path::Combine(GetApplicationFolder(applicationId), L"log");
}

wstring RunLayoutSpecification::GetApplicationTempFolder(
    wstring const & applicationId) const
{
    return Path::Combine(GetApplicationFolder(applicationId), L"temp");
}

wstring RunLayoutSpecification::GetApplicationSettingsFolder(
    wstring const & applicationId) const
{
    return Path::Combine(GetApplicationFolder(applicationId), L"settings");
}

wstring RunLayoutSpecification::GetApplicationPackageFile(
    wstring const & applicationId,
    wstring const & applicationRolloutVersion) const
{
    return Path::Combine(
        GetApplicationFolder(applicationId),
        FileNamePattern::GetApplicationPackageFileName(applicationRolloutVersion));
}

wstring RunLayoutSpecification::GetServicePackageFile(
    wstring const & applicationId,
    wstring const & servicePackageName,
    wstring const & servicePackageRolloutVersion) const
{
    return Path::Combine(
        GetApplicationFolder(applicationId),
        FileNamePattern::GetServicePackageFileName(servicePackageName, servicePackageRolloutVersion));
}

wstring RunLayoutSpecification::GetCurrentServicePackageFile(
    wstring const & applicationId,
    wstring const & servicePackageName,
    wstring const & servicePackageActivationId) const
{
    wstring servicePackageFileName(servicePackageName);

    if (!servicePackageActivationId.empty())
    {
        servicePackageFileName.append(L".");
        servicePackageFileName.append(servicePackageActivationId);
    }

    return GetServicePackageFile(
        applicationId,
        servicePackageFileName,
        L"Current");
}

wstring RunLayoutSpecification::GetEndpointDescriptionsFile(
    wstring const & applicationId,
    wstring const & servicePackageName,
    wstring const & servicePackageActivationId) const
{
    wstring endpointDescriptionsFile(servicePackageName);

    if (!servicePackageActivationId.empty())
    {
        endpointDescriptionsFile.append(L".");
        endpointDescriptionsFile.append(servicePackageActivationId);
    }

    endpointDescriptionsFile.append(L".Endpoints.txt");

    return Path::Combine(
        GetApplicationFolder(applicationId),
        endpointDescriptionsFile);
}


wstring RunLayoutSpecification::GetPrincipalSIDsFile(
    wstring const & applicationId) const
{
    wstring sidsFile = L"Application.Sids.txt";

    return Path::Combine(
        GetApplicationFolder(applicationId),
        sidsFile);
}

wstring RunLayoutSpecification::GetServiceManifestFile(
    wstring const & applicationId,
    wstring const & serviceManifestName,
    wstring const & serviceManifestVersion) const
{
    return Path::Combine(
        GetApplicationFolder(applicationId),
        FileNamePattern::GetServiceManifestFileName(serviceManifestName, serviceManifestVersion));
}

wstring RunLayoutSpecification::GetCodePackageFolder(
    wstring const & applicationId,
    wstring const & servicePackageName,
    wstring const & codePackageName,
    wstring const & codePackageVersion,
    bool isShared) const
{
    wstring path = Path::Combine(
        GetApplicationFolder(applicationId),
        FolderNamePattern::GetCodePackageFolderName(
            servicePackageName,
            codePackageName,
            codePackageVersion));

    if (isShared)
    {
        path.append(RunLayoutSpecification::SharedFolderSuffix);
    }
    return path;
}

wstring RunLayoutSpecification::GetConfigPackageFolder(
    wstring const & applicationId,
    wstring const & servicePackageName,
    wstring const & configPackageName,
    wstring const & configPackageVersion,
    bool isShared) const
{
    wstring path = Path::Combine(
        GetApplicationFolder(applicationId),
        FolderNamePattern::GetConfigPackageFolderName(
            servicePackageName,
            configPackageName,
            configPackageVersion));

    if (isShared)
    {
        return path.append(RunLayoutSpecification::SharedFolderSuffix);
    }
    return path;
}

wstring RunLayoutSpecification::GetDataPackageFolder(
    wstring const & applicationId,
    wstring const & servicePackageName,
    wstring const & dataPackageName,
    wstring const & dataPackageVersion,
    bool isShared) const
{
    wstring path = Path::Combine(
        GetApplicationFolder(applicationId),
        FolderNamePattern::GetDataPackageFolderName(
            servicePackageName,
            dataPackageName,
            dataPackageVersion));

    if (isShared)
    {
        path.append(RunLayoutSpecification::SharedFolderSuffix);
    }
    return path;
}

wstring RunLayoutSpecification::GetSettingsFile(
    wstring const & configPackageFolder) const
{
    return Path::Combine(configPackageFolder, *(FileNamePattern::SettingsFileName));
}

wstring RunLayoutSpecification::GetSubPackageArchiveFile(
    wstring const & packageFolder) const
{
    return ImageModelUtility::GetSubPackageArchiveFileName(packageFolder);
}

File: src/prod/src/managed/Api/src/System/Fabric/Common/ImageModel/RunLayoutSpecification.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Common.ImageModel
{
    using System;
    using System.Fabric.Interop;
    using System.Reflection;

    internal class RunLayoutSpecification
    {
        private NativeImageModel.IFabricRunLayoutSpecification nativeRunLayout;

        private RunLayoutSpecification(NativeImageModel.IFabricRunLayoutSpecification nativeRunLayout)
        {
            this.nativeRunLayout = nativeRunLayout;
        }

        public static RunLayoutSpecification Create()
        {
            return Utility.WrapNativeSyncInvokeInMTA<RunLayoutSpecification>(RunLayoutSpecification.CreateHelper, "RunLayoutSpecification.Create");
        }

        public void SetRoot(string value)
        {
            Utility.WrapNativeSyncInvokeInMTA(() => this.SetRootHelper(value), "StoreLayoutSpecification.SetRoot");
        }

        public string GetApplicationLogFolder(
            string applicationId)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetApplicationLogFolderHelper(applicationId), "RunLayoutSpecification.GetApplicationLogFolder");
        }

        public string GetApplicationPackageFile(
            string applicationId,
            string applicationRolloutVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetApplicationPackageFileHelper(applicationId, applicationRolloutVersion), "RunLayoutSpecification.GetApplicationPackageFile");
        }

        public string GetServicePackageFile(
            string applicationId,
            string servicePackageName,
            string servicePackageRolloutVersion)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetServicePackageFileHelper(applicationId, servicePackageName, servicePackageRolloutVersion), "RunLayoutSpecification.GetServicePackageFile");
        }

        public string GetDataPackageFolder(
            string applicationId, 
            string servicePackageName, 
            string dataPackageName, 
            string dataPackageVersion,
            bool isSharedPackage)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetDataPackageFolderHelper(applicationId, servicePackageName, dataPackageName, dataPackageVersion, isSharedPackage), "RunLayoutSpecification.GetDataPackageFolder");
        }

        public string GetCodePackageFolder(
            string applicationId,
            string servicePackageName,
            string codePackageName,
            string codePackageVersion,
            bool isSharedPackage)
        {
            return Utility.WrapNativeSyncInvokeInMTA(() => this.GetCodePackageFolderHelper(applicationId, servicePackageName, codePackageName, codePackageVersion, isSharedPackage), "RunLayoutSpecification.GetCodePackageFolder");
        }

        #region Interop code

        private static RunLayoutSpecification CreateHelper()
        {
            Guid iid = typeof(NativeImageModel.IFabricRunLayoutSpecification).GetTypeInfo().GUID;
            NativeImageModel.IFabricRunLayoutSpecification newNativeStoreLayout = null;

            newNativeStoreLayout = NativeImageModel.FabricCreateRunLayoutSpecification(ref iid);

            return new RunLayoutSpecification(newNativeStoreLayout);
        }

        private void SetRootHelper(string value)
        {
            using (var pin = new PinBlittable(value))
            {
                this.nativeRunLayout.SetRoot(pin.AddrOfPinnedObject());
            }
        }

        private string GetApplicationLogFolderHelper(
            string applicationId)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeRunLayout.GetApplicationLogFolder(
                    pin.AddBlittable(applicationId));

                return StringResult.FromNative(result);
            }
        }

        private string GetApplicationPackageFileHelper(
            string applicationId,
            string applicationRolloutVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeRunLayout.GetApplicationPackageFile(
                    pin.AddBlittable(applicationId),
                    pin.AddBlittable(applicationRolloutVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetServicePackageFileHelper(
            string applicationId,
            string servicePackageName,
            string servicePackageRolloutVersion)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeRunLayout.GetServicePackageFile(
                    pin.AddBlittable(applicationId),
                    pin.AddBlittable(servicePackageName),
                    pin.AddBlittable(servicePackageRolloutVersion));

                return StringResult.FromNative(result);
            }
        }

        private string GetDataPackageFolderHelper(
            string applicationId,
            string servicePackageName,
            string dataPackageName,
            string dataPackageVersion,
            bool isSharedPackage)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeRunLayout.GetDataPackageFolder(
                    pin.AddBlittable(applicationId),
                    pin.AddBlittable(servicePackageName),
                    pin.AddBlittable(dataPackageName),
                    pin.AddBlittable(dataPackageVersion),
                    NativeTypes.ToBOOLEAN(isSharedPackage));

                return StringResult.FromNative(result);
            }
        }

        private string GetCodePackageFolderHelper(
            string applicationId,
            string servicePackageName,
            string codePackageName,
            string codePackageVersion,
            bool isSharedPackage)
        {
            using (var pin = new PinCollection())
            {
                var result = this.nativeRunLayout.GetCodePackageFolder(
                    pin.AddBlittable(applicationId),
                    pin.AddBlittable(servicePackageName),
                    pin.AddBlittable(codePackageName),
                    pin.AddBlittable(codePackageVersion),
                    NativeTypes.ToBOOLEAN(isSharedPackage));

                return StringResult.FromNative(result);
            }
        }

        #endregion
    }
}